[Patterns]: Add new `Header Centered Search` pattern

albarin opened the pull request:
This PR implements the `Header Centered Search` pattern.

Fixes https://github.com/woocommerce/woocommerce-blocks/issues/9035

### Screenshots

#### Design
![Centered Search](https://user-images.githubusercontent.com/186112/232716016-cf9ec8de-b8ad-4d35-a081-cd9112daa535.png)

#### Implementation
<img width="1362" alt="Screenshot 2023-04-18 at 10 15 38" src="https://user-images.githubusercontent.com/186112/232715398-99c12882-8ffb-40d5-b5be-7b66460e64b3.png">


### Testing
#### User-Facing Testing

1. Create a new page or post
2. Make sure the `WooCommerce Header Centered Search` pattern appears under the WooCommerce category dropdown.
3. Insert in and make sure it shows as expected on the design.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog
> Add new `Header Centered Search` pattern

albarin committed:
Add new `Header Centered Search` pattern

kmanijak commented on patterns/header-centered-search-pattern.php:
```diff
@@ -0,0 +1,73 @@
+
+			<!-- wp:spacer -->
+			<div style="height:100px" aria-hidden="true" class="wp-block-spacer"></div>
+			<!-- /wp:spacer -->
```
What's the purpose of multiple Spacer blocks between Search and Customer Account? Could they be removed? I did that in Editor and I think the layout stayed the same.

albarin replied:
Yeah, it's not a good solution, I wanted to add some margin between the search block and the customer account block. I've added support for `margin` on the `Customer Account` block, and added some left margin to it.

kmanijak reviewed (commented):
I can find the pattern and apply it, however the search seems to by squeezed:
| Editor | <img width="2265" alt="image" src="https://user-images.githubusercontent.com/20098064/232741775-78569d12-86ae-48f9-ba2b-49ed08a45879.png"> |
| --- | --- |
| Frontend | <img width="2546" alt="image" src="https://user-images.githubusercontent.com/20098064/232741943-d3a15a3c-b759-4e89-81ff-9c9176a0d336.png"> |

As you can see the placeholder doesn't show up fully even though there's a space for it. I tried to use the width control, but that doesn't help and seems buggy (I don't consider it a part of this PR though, just wanted to show how it behaves).
https://user-images.githubusercontent.com/20098064/232742699-9334cee0-030c-4e2f-9c39-037e0f710744.mov

Do you think there are additional steps/conditions that I should follow/satisfy?

albarin committed:
Add margin support to the customer account block

albarin committed:
Add margin to the customer account to separate it from the search bar

albarin committed:
Update margin

albarin committed:
Add with, fix margins

albarin commented:
@kmanijak yep, that's weird, I've pushed some updates, I'm trying with a width in `px`, wdyt? 🤔

kmanijak commented on patterns/header-centered-search-pattern.php:
```diff
@@ -0,0 +1,54 @@
+
+		<!-- wp:group {"style":{"spacing":{"blockGap":"8px"},"typography":{"fontSize":"13px"}},"layout":{"type":"flex","flexWrap":"nowrap","justifyContent":"right"}} -->
+		<div class="wp-block-group" style="font-size:13px">
+			<!-- wp:group {"style":{"spacing":{"blockGap":"16px"}},"layout":{"type":"flex","flexWrap":"nowrap"}} -->
+			<div class="wp-block-group">
+				<!-- wp:separator {"className":"is-style-wide"} -->
+				<hr class="wp-block-separator has-alpha-channel-opacity is-style-wide" />
+				<!-- /wp:separator -->
+			</div>
+			<!-- /wp:group -->
```
The only thing I found, was that this Row with Separator inside doesn't influence the look of the template. Could they be removed or there's some other reason that I missed?

| With | <img width="1452" alt="Screen Shot 2023-04-19 at 08 58 16 AM" src="https://user-images.githubusercontent.com/20098064/232992119-1dd649bb-afb7-439f-ae6b-27aa939149ef.png"> |
| --- | --- |
| Without | <img width="1453" alt="image" src="https://user-images.githubusercontent.com/20098064/232991942-bb4912d0-5e5d-4405-8a10-3d427f95a33d.png"> |

albarin replied:
Thanks, removed!

kmanijak reviewed (approved):
That tests well! I left one comment about a Row with Separator, but pre-approving! 👍

kmanijak commented:
> @kmanijak yep, that's weird, I've pushed some updates, I'm trying with a width in px, wdyt? 🤔

I like the idea, because there's a handle on the right and merchant can easily adjust the width in case they need it, so for me, it's a win-win:
<img width="392" alt="image" src="https://user-images.githubusercontent.com/20098064/232993037-f76a1b55-d454-49b0-a3fc-e0e33244591e.png">

albarin committed:
Remove unnecessary separator

albarin commented:
@shaunandrews this is ready for you to take a look, also any suggestions for the pattern title? Thanks! 😄

shaunandrews commented:
<img width="367" alt="CleanShot 2023-04-19 at 12 06 47@2x" src="https://user-images.githubusercontent.com/191598/233135018-eb8c3a65-4aea-4614-baeb-fc25131c19d8.png">


I think the block spacing between the logo and title could be reduced down by half to 20px.

--

<img width="390" alt="CleanShot 2023-04-19 at 12 05 53@2x" src="https://user-images.githubusercontent.com/191598/233134893-41b3c38e-e2f5-42c1-834e-ec3100b638e9.png">

The search looks strange to me. Almost like it's not rendering the right CSS. This is with the Tsubaki theme.

--

<img width="144" alt="CleanShot 2023-04-19 at 12 07 10@2x" src="https://user-images.githubusercontent.com/191598/233135209-af1f1175-7b7f-41b4-b43e-bb456243bfc0.png">

On the Navigation block I changed the Font Appearance to "Normal" and found that things look much better that the bolded links I saw at first. Is this is theme thing?

--

For the Search placeholder lets reduce it down to simply "Search our store".

shaunandrews commented:
Oh, and for a name, how about "Centered Menu with Search"?

albarin committed:
Updates after review

albarin commented:
@shaunandrews I addressed all your feedback.
I think the solution I did for the search input border is maybe too hacky, so let's discuss it with the team before merging.

albarin committed:
Remove ref

albarin committed:
Remove separator

albarin committed:
Bring back separator and use product search

albarin committed:
Merge branch 'trunk' into 9035/header-centered-search-pattern

thealexandrelara commented:
WooCommerce Blocks 10.1.0 will be released soon. I'm bumping this PR to the next milestone

shaunandrews commented:
<img width="1376" alt="CleanShot 2023-04-24 at 12 27 00@2x" src="https://user-images.githubusercontent.com/191598/234058503-1070962c-fa88-47c7-8ae6-6bfd89a0fcb0.png">

While clever, I don't think using the HTML block is the right solution; It adds the content to the canvas in a confusing way.

The border comes from the block itself: https://github.com/WordPress/gutenberg/blob/fd98542335ce1a2eb0517b9e4f285d750b4c4645/packages/block-library/src/search/style.scss#L59

Its unclear to me why, though.

I think a potential solution here is to update the Tsubaki theme or the Search block.

albarin committed:
Merge branch 'trunk' into 9035/header-centered-search-pattern

albarin committed:
Remove html block and fix search border

albarin commented:
> While clever, I don't think using the HTML block is the right solution; It adds the content to the canvas in a confusing way.

@shaunandrews Agree, I pushed another solution I thought today: instead of using `Button inside` use `Button outside` and adding the border outside with a stack element, I think it works. Let me know what you think 😄
